rfcs/text/0013_saved_object_migrations.md

```diff
@@ -0,0 +1,527 @@
+migrations are idempotent: 
+
+| Idempotent migrations |Description                                                |
+| --------------------- | --------------------------------------------------------- |
+| Yes                   | Idempotent migrations performed without coordination      |
+| No                    | Single node migrations coordinated through a lease / lock |
+
+Idempotent migrations don't require coordination making the algorithm
+significantly less complex and will never require manual intervention to
+retry. We, therefore, prefer this solution, even though it introduces
+restrictions on migrations (4.3.1.1). For other alternatives that were
+considered see section [(5)](#5-alternatives).
+
+## 4.3.1 Idempotent migrations performed without coordination
+
+The migration system can be said to be idempotent if the same results are
+produced whether the migration was run once or multiple times. This property
+should hold even if new (up to date) writes occur in between migration runs
+which introduces the following restrictions:
+
+### 4.3.1.1 Restrictions
+
+1. All document transforms need to be deterministic, that is a document
+   transform will always return the same result for the same set of inputs. 
+2. It should always be possible to construct the exact set of inputs required
+   for (1) at any point during the migration process (before, during, after). 
+
+Although these restrictions require significant changes, it does not prevent
+known upcoming migrations such as [sharing saved-objects in multiple spaces](https://github.com/elastic/kibana/issues/27004) or [splitting a saved
+object into multiple child
+documents](https://github.com/elastic/kibana/issues/26602). To ensure that
+these migrations are idempotent, they will have to generate new saved object
+id's deterministically with e.g. UUIDv5.
+
+### 4.3.1.2 Algorithm
+
+1. On startup, find the index to which the `.kibana` alias points,
+   `.kibana_n`, and check if any documents have an outdated `migrationVersion`
+   indicating that a migration needs to be performed. 
+2. Create `.kibana_n+1`, if it already exists, ignore the error and continue.
+3. Migrate documents by reading from `.kibana_n` and creating documents in
+   `.kibana_n+1`. If a document already exists, don't overwrite it and ignore.
+4. Move the `.kibana` index to point to `.kibana_n+1` (ignore if it has
+   already been moved). 
+
+Together with the limitations this algorithm ensures that migrations are
+idempotent. If two nodes are started simultaneously, both of them will start
+writing into `.kibana_n+1` but because migrations are idempotent, it doesn’t
+matter which node’s writes win.
+
+> Note: This algorithm doesn't support mixed Kibana versions attempting to
+> perform a migration at the same time. E.g. if a v7.8.0 and a v7.9.0 node
+> attempts to simultaneously migrate a 7.6.0 index the result will be an
+> inconsistent state. This is an unlikely scenario in practise and therefore
```

Comment:
> This is an unlikely scenario in practise and therefore an acceptable limitation

This limitation makes sense but I just want to cast a vote for fixing this problem sooner rather than later:

For Observability development we have a GCP machine that ingests gigabytes of data per hour into a centralised cloud-based Elasticsearch instance. To avoid having to run a similar setup locally (very resource and storage intensive) developers connect their local Kibana instance to the shared ES instance.

Currently we have big problems with migrations on developers machines interfering with each other. Especially if a developer downgrades their Kibana instance to run a previous version while still connected to the shared ES instance. I can imagine this is not that uncommon for other Kibana devs.

It sounds like this proposal will alleviate the problems we are seeing so I would love for it to happen before 8.x (which could be years away).
Regardless what timeline you end up with, I love that you are working on improving this 👍 👍

Reply:
I should probably use 8.0 instead of 8.x, this will probably not change for the entire 8.x, but the idea is to release it as part of 8.0 so at least that brings the timeline closer :)

Reply:
It does! Thanks for the clarification.